pkg/chaosctl/cmd/completion.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2020 Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"os"
+
+	"github.com/spf13/cobra"
+)
+
+// completionCmd represents the completion command
+var completionCmd = &cobra.Command{
+	Use:   "completion [bash|zsh|fish|powershell]",
+	Short: "Generate completion script",
+	Long: `To load completions:
+
+Bash:
+
+$ source <(chaosctl completion bash)
+
+# To load completions for each session, execute once:
+Linux:
+  $ chaosctl completion bash > /etc/bash_completion.d/chaosctl
+MacOS:
+  $ chaosctl completion bash > /usr/local/etc/bash_completion.d/chaosctl
+
+Zsh:
+
+# If shell completion is not already enabled in your environment you will need
+# to enable it.  You can execute the following once:
+
+$ echo "autoload -U compinit; compinit" >> ~/.zshrc
+
+# To load completions for each session, execute once:
+$ chaosctl completion zsh > "${fpath[1]}/_chaosctl"
+
+# You will need to start a new shell for this setup to take effect.
+
+Fish:
+
+$ chaosctl completion fish | source
+
+# To load completions for each session, execute once:
+$ chaosctl completion fish > ~/.config/fish/completions/chaosctl.fish
+`,
+	DisableFlagsInUseLine: true,
+	ValidArgs:             []string{"bash", "zsh", "fish", "powershell"},
+	Args:                  cobra.ExactValidArgs(1),
+	Run: func(cmd *cobra.Command, args []string) {
+		switch args[0] {
+		case "bash":
+			cmd.Root().GenBashCompletion(os.Stdout)
+		case "zsh":
+			cmd.Root().GenZshCompletion(os.Stdout)
```

Comment:
Cobra's support for `zsh` completion is very limited and usually doesn't work well. It would be good if we can reuse the code from `kubectl`. https://github.com/kubernetes/kubectl/blob/master/pkg/cmd/completion/completion.go#L145

Reply:
Don't be afraid of the hard coding scripts :). The change should be easy. Just rename `kubectl` to `chaosctl` would be fine.

Reply:
Hi @yeya24 Actually cobra's support for zsh just got updated after https://github.com/spf13/cobra/pull/1070, and I think kubectl community is also working on this update and migration on https://github.com/kubernetes/kubectl/issues/882.

Reply:
A screenshot for this in zsh:
```
 ~  chaosctl debug networkchaos -n kube-
kube-node-lease  kube-public      kube-system
```

Reply:
It just we need to generate clientset for chaos CRDs to support CRUD(I think controller-gen could do this), to make auto-complete for chaos work

Reply:
Cool! That's perfect. Thanks for letting me know.